### DISCUSSION TITLE
node storage without longhorn

### DISCUSSION THREAD

**Discussion by mkreuzmayr** [mysticaltech/terraform-hcloud-kube-hetzner#465] [Q&A] [answered]
Is it possible to store a PVC on node storage (for a database) without using longhorn?
Longhorn would be overkill and to resource heavy for my use case.

**Comment by liaralabs** [comment, marked as answer]
Sure, you can use hostPath volumes and you can even use something like a [host path provisioner](https://github.com/rancher/local-path-provisioner) to help ease the pain of creating such mounts.

You'll generally need to pair hostPath volumes with a nodeSelector though as any data stored in hostPaths won't be accessible to other nodes if you have more than one instance. I'm not sure if that provisioner handles the nodeselection/taints automatically, but it's something to be aware of.

Alternatively, you can provision hetzner volumes for this task and use the hetzner CSI to automatically provision your requests and attach the volume to the server where the pod intends to mount the data. These volumes give you a bit more flexibility in that they can be attached to other nodes if necessary. Thus your data becomes a bit less locked to a specific individual server, though the performance of these volumes is significantly worse than the local NVMe storage. These volumes can only be attached to a single server though and can only be mounted RWO. If using it with a deployment, you'll have to use the "Recreate" rather than rolling strategy for rolling out new pods as your rollouts will just be stuck waiting to bind to a volume that will never be released. There will be ~1 minute of downtime as the volume/pod moves between servers.

If you want an HA solution that will give you the ability to attach the pod to whatever server, whenever, wherever, use rolling strategies and rollout new deployments or swap statefulsets between nodes without (or with minimal) downtime then you need a backend which is capable of being mounted RWX. Longhorn is the "easiest" of these and it only gets more complicated and convoluted from there (rook ceph, openebs, etc).

**Reply by mysticaltech** [reply]
@liaralabs Great answer.

 @mkreuzmayr If I were you I would still use Longhorn for max flexibility, and data duplication which can be super useful if bad scenarios.

**Reply by mkreuzmayr** [reply]
@liaralabs @mysticaltech Thanks for your detailed answer. I have got 3 CPX31 running a HA keycloak instance with a zalando/patroni database, would i be a big performance hog if I add longshorn for database PVC management? Or would I need to upgrade servers for that?

**Reply by mysticaltech** [reply]
I do not think so, with CPX31 it should be enough IMHO.
